src/main/java/org/traccar/protocol/Xexun2ProtocolEncoder.java

```diff
@@ -0,0 +1,121 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.traccar.protocol;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import org.traccar.BaseProtocolEncoder;
+import org.traccar.helper.DataConverter;
+import org.traccar.model.Command;
+import org.traccar.Protocol;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+
+public class Xexun2ProtocolEncoder extends BaseProtocolEncoder {
+
+    public Xexun2ProtocolEncoder(Protocol protocol) {
+        super(protocol);
+    }
+
+    private static ByteBuf encodeFrame(ByteBuf buf) {
+        int bufLength = buf.readableBytes();
+        if (bufLength < 5) {
+            return null;
+        }
+
+        ByteBuf result = Unpooled.buffer();
+
+        result.writeBytes(buf.readBytes(2));
+
+        while (buf.readerIndex() < bufLength - 2) {
+            int b = buf.readUnsignedByte();
+            if (b == 0xfa && buf.isReadable() && buf.getUnsignedByte(buf.readerIndex()) == 0xaf) {
+                buf.readUnsignedByte();
+                result.writeByte(0xfb);
+                result.writeByte(0xbf);
+                result.writeByte(0x01);
+            } else if (b == 0xfb && buf.isReadable() && buf.getUnsignedByte(buf.readerIndex()) == 0xbf) {
+                buf.readUnsignedByte();
+                result.writeByte(0xfb);
+                result.writeByte(0xbf);
+                result.writeByte(0x02);
+            } else {
+                result.writeByte(b);
+            }
+        }
+        result.writeBytes(buf.readBytes(2));
+
+        return result;
+    }
+
+    private static int checksum(byte[] data)
+    {
+        int sum = 0;
+        int len = data.length;
+        for (int j = 0; len > 1; len--) {
+            sum += data[j++] & 0xff;
+            if ((sum & 0x80000000) > 0) {
+                sum = (sum & 0xffff) + (sum >> 16);
+            }
+        }
+        if (len == 1) {
+            sum += data[data.length - 1] & 0xff;
+        }
+        while ((sum >> 16) > 0) {
+            sum = (sum & 0xffff) + sum >> 16;
+        }
+        sum = (sum == 0xffff) ? sum & 0xffff : (~sum) & 0xffff;
+        return sum;
+    }
+
+
+    private static ByteBuf encodeContent(String uniqueId, String content) {
+        ByteBuf buf = Unpooled.buffer();
+
+        byte[] message = content.getBytes();
+
+        buf.writeShort(0xFAAF);
+        buf.writeShort(0x0007);
+        buf.writeShort(0x0001);
```

Comment:
What are these?

Reply:
0x0001 is Serial Number for message. With messages coming from device it increases by 1 each message sent to server.

For messages sent to device you could increase each time a message is sent so that you can directly match ack to message sent. Would this be possible with traccar? Is there an enumerated property for Command passed to encodeCommand()?

I have had no issues using 0x0001 for every command sent to X06 device.

Reply:
Then write it like this:
```
buf.writeShort(1); // index
```

What about the other one?

Reply:
message id 0x07 = remote control command

Reply:
You should reuse the type from `Xexun2ProtocolDecoder`.